### PR TITLE
[feature] add support of a runc sandboxer

### PR DESCRIPTION
**Pull request by abel-von**
A remote runc sandboxer which will run a process in seperate namespaces and this process will serve the Task API by a unix domain socket, and this process will also be the process to carry the common namespaces of containers in the same pod. so that we don't have a pause process, the "shim" process will act as the "pause" process

### CODE REVIEW

**Comment by xiaods** [comment]
any update on it?

**Comment by abel-von** [comment]
Maybe we will have a community meeting next week to discuss the design. @xiaods

**Comment by xiaods** [comment]
> Maybe we will have a community meeting next week to discuss the design. @xiaods

any update on it?

**Comment by xiaods** [comment]
> Maybe we will have a community meeting next week to discuss the design. @xiaods

any update on this community meeting?

**Comment by Burning1020** [comment]
> > Maybe we will have a community meeting next week to discuss the design. @xiaods
> 
> 
> 
> any update on this community meeting?

Sorry, Abel has been working for the Kubecon China, should we have it after this vocation? @abel-von Do you have time?

**Comment by xiaods** [comment]
> > > Maybe we will have a community meeting next week to discuss the design. @xiaods
> > 
> > 
> > any update on this community meeting?
> 
> Sorry, Abel has been working for the Kubecon China, should we have it after this vocation? @abel-von Do you have time?

I just want to learn about the latest project progress and it doesn't matter about the time. I also need more time to understand this project, so there is no need to worry about time pressure. Thank you.

**Comment by xiaods** [comment]
@abel-von  if the runc sandboxer feature is implemented, can I replace the containerd with the kuasar component?

**Comment by xiaods** [comment]
any update?

File: runc/src/exitsnoop.c [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+};
+
+BPF_PERF_OUTPUT(events);
+
+//TRACEPOINT_PROBE(sched, sched_process_exit)
+int sched_process_exit_handler(struct tracepoint__sched__sched_process_exit *args)
+{
+    struct task_struct *task = (typeof(task)) bpf_get_current_task();
+    struct process_exit_data_t data = {};
+    //data.start_time = PROCESS_START_TIME_NS,
```
Comment (xiaods) [inline runc/src/exitsnoop.c:21]: What does this line of code mean? If it's not necessary, it can be removed.

**Comment by Burning1020** [comment]
> any update?

The code is basically complete but still needs some time.

**Review by Burning1020 — COMMENTED** [review]
File Header and CI are needed, the other comment could be resolved in follow up.

**Comment by Burning1020** [comment]
We also have to provide a `make` command for building.

**Comment by flyflypeng** [comment]
> We will follow this design to make a runc sandboxer image

I think it's necessary to update this architecture diagram corresponding to the runc sandboxer in the `README.md` and provide additional details about the design."

**Comment by flyflypeng** [comment]
> We will follow this design to make a runc sandboxer image

I think this arch

File: runc/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+                SaFlags::empty(),
+                SigSet::empty(),
+            );
+            unsafe {
+                sigaction(SIGCHLD, &sig_action).unwrap();
+            }
+            loop {
+                let buffer = read_count(reqr, 512).unwrap();
+                let id = String::from_utf8_lossy(&buffer[0..64]).to_string();
+                let mut zero_index = 64;
```
Comment (flyflypeng) [inline runc/src/main.rs:101]: what's the meaning of the 64, a static const maybe suitable

File: runc/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+                SigHandler::Handler(sandbox_parent_handle_signals),
+                SaFlags::empty(),
+                SigSet::empty(),
+            );
+            unsafe {
+                sigaction(SIGCHLD, &sig_action).unwrap();
+            }
+            loop {
+                let buffer = read_count(reqr, 512).unwrap();
+                let id = String::from_utf8_lossy(&buffer[0..64]).to_string();
```
Comment (flyflypeng) [inline runc/src/main.rs:100]: If `id` means runc sandbox instance id, why not use `String::from_utf8`

File: runc/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+        ForkResult::Parent { child } => {
+            debug!("forked process {} for the sandbox parent", child);
+            close(reqr).unwrap_or_default();
+            close(respw).unwrap_or_default();
+        }
+        ForkResult::Child => {
+            close(reqw).unwrap_or_default();
+            close(respr).unwrap_or_default();
+            prctl::set_child_subreaper(true).unwrap();
+            let comm = format!("[sandbox-parent]");
```
[the remark this answers is not in the document]
Comment (abel-von) [reply runc/src/main.rs]: I just want to to highlight the process as a special process, do you have any good suggestion?

File: runc/src/sandbox.rs [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+                    Some("proc"),
+                    MsFlags::MS_NOSUID | MsFlags::MS_NOEXEC | MsFlags::MS_NODEV,
+                    None::<&str>,
+                )?;
+                Ok(())
+            });
+        }
+        let log_pipe = format!("{}/log", self.base_dir);
+        unsafe {
+            mkfifo(&*log_pipe, Mode::from_bits_unchecked(0700))
```
[the remark this answers is not in the document]
Comment (abel-von) [reply runc/src/sandbox.rs]: codes removed

File: runc/src/main.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (flyflypeng) [reply runc/src/main.rs]: I think it's sufficient to distinguish using the unique process name directly, there's no need to add square brackets.

File: runc/src/main.rs (hunk 3 above)
Comment (flyflypeng) [reply runc/src/main.rs:99]: I agree.

**Review by Burning1020 — APPROVED** [review]
lgtm

**Pull request merged**